server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java

```diff
@@ -73,12 +83,124 @@ public void useAggregatedNodeLevelResponses(boolean useAggregatedNodeLevelRespon
         this.useAggregatedNodeLevelResponses = useAggregatedNodeLevelResponses;
     }
 
+    /**
+     * Get the names of requested metrics, excluding indices, which are
+     * handled separately.
+     */
+    public Set<String> requestedMetrics() {
+        return new HashSet<>(requestedMetrics);
+    }
+
+    /**
+     * Add subMetric
+     */
+    public ClusterStatsRequest addMetric(String subMetric) {
+        if (SubMetrics.allSubMetrics().contains(subMetric) == false) {
+            throw new IllegalStateException("Used an illegal subMetric: " + subMetric);
+        }
+        requestedMetrics.add(subMetric);
+        return this;
+    }
+
+    public void clearMetrics() {
+        requestedMetrics.clear();
+    }
+
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         if (out.getVersion().onOrAfter(Version.V_2_16_0)) {
             out.writeOptionalBoolean(useAggregatedNodeLevelResponses);
         }
+        if (out.getVersion().onOrAfter(Version.V_3_0_0)) {
+            out.writeStringArray(requestedMetrics.toArray(new String[0]));
+        }
+    }
+
+    /**
+     * An enumeration of the "core" sections of metrics that may be requested
+     * from the cluster stats endpoint.
+     */
+    @PublicApi(since = "3.0.0")
+    public enum Metric {
+        INDICES("indices"),
+        NODES("nodes");
+
+        private String metricName;
+
+        Metric(String name) {
+            this.metricName = name;
+        }
+
+        public String metricName() {
+            return this.metricName;
+        }
+
+        public Set<String> getSubMetrics() {
+            return SubMetrics.allSubMetrics(this);
+        }
     }
 
+    /**
+     * An enumeration of the "core" sections of sub metrics that may be requested
+     * from the cluster stats endpoint.
+     */
+    public enum SubMetrics {
+        SHARDS("shards", Metric.INDICES),
+        DOCS("docs", Metric.INDICES),
+        STORE("store", Metric.INDICES),
+        FIELDDATA("fielddata", Metric.INDICES),
+        QUERY_CACHE("query_cache", Metric.INDICES),
+        COMPLETION("completion", Metric.INDICES),
+        SEGMENTS("segments", Metric.INDICES),
+        ANALYSIS("analysis", Metric.INDICES),
+        MAPPINGS("mappings", Metric.INDICES),
+
+        OS("os", Metric.NODES),
+        PROCESS("process", Metric.NODES),
+        JVM("jvm", Metric.NODES),
+        FS("fs", Metric.NODES),
```

Comment:
The nodes stats define these as Metrics although this is reported per-node. Should we keep it consistent. The 2nd metric path is defined to be a `index-metric`
https://opensearch.org/docs/latest/api-reference/nodes-apis/nodes-stats/#path-parameters

Reply:
Discussed offline on this and in this PR we will have the metric filters only along with node filter.
The URI would be `/_cluster/stats/nodes/{nodeId}/{metric}/{index_metric}`